connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/internal/columnNodeSupport.scala

```diff
@@ -212,9 +221,68 @@ object ColumnNodeToProtoConverter extends (ColumnNode => proto.Expression) {
   }
 }
 
+object ColumnNodeUtil {
+  def collectReferences(nodes: Seq[ColumnNode]): Seq[proto.Relation] = {
+    nodes.flatMap {
+      case UnresolvedFunction(_, arguments, _, _, _, _) =>
+        collectReferences(arguments)
+      case Alias(child, _, _, _) =>
+        collectReferences(Seq(child))
+      case Cast(child, _, _, _) =>
+        collectReferences(Seq(child))
+      case Window(windowFunction, windowSpec, _) =>
+        collectReferences(
+          Seq(windowFunction) ++ windowSpec.partitionColumns ++ windowSpec.sortColumns) ++
+          windowSpec.frame.toSeq.flatMap { frame =>
+            collectReferences(frame.lower) ++
+              collectReferences(frame.upper)
+          }
+      case UnresolvedExtractValue(child, extraction, _) =>
+        collectReferences(Seq(child, extraction))
+      case UpdateFields(structExpression, _, valueExpression, _) =>
+        collectReferences(Seq(structExpression) ++ valueExpression.toSeq)
+      case LambdaFunction(function, _, _) =>
+        collectReferences(Seq(function))
+      case InvokeInlineUserDefinedFunction(_, args, _, _) =>
+        collectReferences(args)
+      case CaseWhenOtherwise(branches, otherwise, _) =>
+        collectReferences(branches.flatMap { case (condition, value) =>
+          Seq(condition, value)
+        } ++ otherwise.toSeq)
+      case LazyExpression(child, _) =>
+        collectReferences(Seq(child))
+      case SubqueryExpressionNode(query, _, _) =>
+        Seq(query)
+      case _ => Seq.empty
+    }
+  }
+
+  private def collectReferences(boundary: WindowFrame.FrameBoundary): Seq[proto.Relation] = {
+    boundary match {
+      case WindowFrame.Value(value) => collectReferences(Seq(value))
+      case _ => Seq.empty
+    }
+  }
+}
+
 case class ProtoColumnNode(
     expr: proto.Expression,
     override val origin: Origin = CurrentOrigin.get)
     extends ColumnNode {
   override def sql: String = expr.toString
 }
+
+sealed trait SubqueryType
+
+object SubqueryType {
+  case object SCALAR extends SubqueryType
+  case object EXISTS extends SubqueryType
+}
+
+case class SubqueryExpressionNode(
+    relation: proto.Relation,
+    subqueryType: SubqueryType,
+    override val origin: Origin = CurrentOrigin.get)
+    extends ColumnNode {
+  override def sql: String = s"$subqueryType ($relation)"
```

Comment:
For SCALAR this does not produce valid SQL. That should be `()`. However `relation.toString` does not produce valid SQL either...

Reply:
Sure, I'll update it for `SCALAR`.
What would you recommend about the `relation.toString` part?

Reply:
I don't think I have anything to recommend there. Rendering SQL for connect plans is not trivial. The purpose of sql is to provide a sql like string, not an actually working piece of sql.